File: src/binary_search.c
// Program for binary search method

#include <stdio.h>

int binarySearch(int arr[], int n, int se);

void main()
{
    int arr[100], n, i, se, pos;

    // Input
    printf("Enter the number of elements: ");
    scanf("%d", &n);

    printf("Enter the elements:\n");
    for (i = 0; i < n; i++)
        scanf("%d", &arr[i]);

    printf("Enter the search element: ");
    scanf("%d", &se);

    printf("The array elements are:\n");
    for (i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");

    // Process
    pos = binarySearch(arr, n, se);

    // Output
    if (pos > 0)
        printf("Element found at location %d", pos);
    else
        printf("Search unsuccessful");
    printf("\n");
}

int binarySearch(int arr[], int n, int se)
{
    int loc = -1, low = 0, high = n - 1, mid;

    // Process - binary search logic
    while (low <= high)
    {
        mid = (low + high) / 2;

        if (se == arr[mid]) // element found
        {
            loc = mid;
            break;
        }
        else if (se < arr[mid]) // search left half
            high = mid - 1;
        else // search right half
            low = mid + 1;
    }
    return loc + 1;
}

File: python_prgms/src/binary_search.py
# Python script to perform binary search

def binary_search(arr, target):
    left = 0
    right = len(arr) - 1

    while left <= right:
        mid = (left + right) // 2

        if arr[mid] == target:
            return mid + 1 # 1-based index
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1

def main():
    arr = sorted([int(i) for i in input("Enter numbers: ").split()]) # Sorts the input array
    target = int(input("Enter number to search: "))

    # Print array elements
    print("The sorted array elements are:")
    print(*arr) # No need to loop while printing array elements. Use * to avoid [sqaure brackets around numbers]

    result = binary_search(arr, target)

    if result != -1:
        print(f"Found at position {result}")
    else:
        print("Not found")

if __name__ == "__main__":
    main()
